app/controllers/incomeSources/cease/BusinessEndDateController.scala

```diff
@@ -50,98 +52,153 @@ class BusinessEndDateController @Inject()(val authenticate: AuthenticationPredic
+              id = id)
         }
   }
 
-  def handleSubmitRequest(id: String, isAgent: Boolean)(implicit user: MtdItUser[_]): Future[Result] = {
-
-    val (postAction, backAction, redirectAction) = {
-      if (isAgent)
-        (routes.BusinessEndDateController.submitAgent _,
-          routes.CeaseIncomeSourceController.showAgent(),
-          routes.CheckCeaseBusinessDetailsController.showAgent())
-      else
-        (routes.BusinessEndDateController.submit _,
-          routes.CeaseIncomeSourceController.show(),
-          routes.CheckCeaseBusinessDetailsController.show())
-    }
+  def handleSubmitRequest(id: Option[String], incomeSourceType: String, isAgent: Boolean)
+                         (implicit user: MtdItUser[_], messages: Messages): Future[Result] = withIncomeSourcesFS {
+    Future {
+      val (backAction, postAction, redirectAction, incomeSourceTypeValue) = getActions(isAgent, incomeSourceType, id)
+      businessEndDateForm.apply(incomeSourceTypeValue, id).bindFromRequest().fold(
 
-    businessEndDateForm.apply(user, id).bindFromRequest().fold(
-      hasErrors => Future.successful(BadRequest(businessEndDate(
-        BusinessEndDateForm = hasErrors,
-        postAction = postAction(id),
-        backUrl = backAction.url,
-        isAgent = isAgent,
-        btaNavPartial = user.btaNavPartial
-      ))),
-      validatedInput =>
-        Future.successful(Redirect(redirectAction)
-          .addingToSession(ceaseBusinessEndDate -> validatedInput.date.toString)
-          .addingToSession(ceaseBusinessIncomeSourceId -> id))
-    )
+        hasErrors => BadRequest(businessEndDate(
+          businessEndDateForm = hasErrors,
+          postAction = postAction,
+          backUrl = backAction.url,
+          isAgent = isAgent,
+          incomeSourceType = incomeSourceTypeValue
+        )(user, messages)),
+
+        validatedInput => {
+          if (incomeSourceTypeValue == SelfEmployment) {
+
+            val incomeSourceId = id.getOrElse(
+              throw new InternalServerException("[BusinessEndDateController][handleSubmitRequest]: Missing income source ID")
+            )
+
+            Redirect(redirectAction)
+              .addingToSession(incomeSourceTypeValue.endDateSessionKey -> validatedInput.date.toString)
+              .addingToSession(ceaseBusinessIncomeSourceId -> incomeSourceId)
+
+          } else {
+
+            Redirect(redirectAction)
+              .addingToSession(incomeSourceTypeValue.endDateSessionKey -> validatedInput.date.toString)
+
+          }
+        })
+    } recover {
+      case ex: Exception =>
+        val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+        Logger("application").error(s"${if (isAgent) "[Agent]"}" +
+          s"[BusinessEndDateController][handleSubmitRequest]: Error getting BusinessEndDate page: ${ex.getMessage}")
+        errorHandler.showInternalServerError()
+    }
   }
```

Comment:
Not really a fan of a massive method like this.. you could pull out the form errors into handleFormErrors, validatedInput into handleFormSuccess, etc. 
Read the functions section here, this block is clearly doing more than one thing:
https://alvinalexander.com/programming/clean-code-quotes-robert-c-martin/

Reply:
Also, there are performance implications of wrapping this entire block in a future in this way, as it will immediately make the execution asynchronous and spill the computation to a new thread (this will not be scalable with thousands of users)

Reply:
sure.

Reply:
@avelx please see https://blog.colinbreck.com/future-successful-blocking-your-future-success/

Reply:
@EdwardDennis  - we are in the async context already, you can verify that we are using same thread ID all along by adding something in nested Future we are using etc
`println(s"ThreadIDA :" + Thread.currentThread().getId )`